florist/api/routes/server/training.py

```diff
@@ -104,3 +117,76 @@ async def start(job_id: str, request: Request) -> JSONResponse:
     except Exception as ex:
         LOGGER.exception(ex)
         return JSONResponse({"error": str(ex)}, status_code=500)
+
+
+def server_training_listener(job: Job, database: Database[Dict[str, Any]]) -> None:
+    """
+    Listen to the Redis' channel that reports updates on the training process of a FL server.
+
+    Keeps consuming updates to the channel until it finds `fit_end` in the server metrics,
+    then closes the job with FINISHED_SUCCESSFULLY and saves both the clients and server's metrics
+    to the job in the database.
+
+    :param job: (Job) The job with the server_uuid to listen to.
+    :param database: (pymongo.database.Database) An instance of the database to save the information
+        into the Job. MUST BE A SYNCHRONOUS DATABASE since this function cannot be marked as async
+        because of limitations with FastAPI's BrackgroundTasks.
+    """
+    LOGGER.info(f"Starting listener for server messages from job {job.id} at channel {job.server_uuid}")
+
+    assert job.server_uuid is not None, "job.server_uuid is None."
+    assert job.redis_host is not None, "job.redis_host is None."
+    assert job.redis_port is not None, "job.redis_port is None."
+
+    # check if training has already finished before start listening
+    server_metrics = get_from_redis(job.server_uuid, job.redis_host, job.redis_port)
+    LOGGER.debug(f"Listener: Current metrics for job {job.id}: {server_metrics}")
+    if server_metrics is not None and "fit_end" in server_metrics:
+        close_job(job, server_metrics, database)
+        return
+
+    subscriber = get_subscriber(job.server_uuid, job.redis_host, job.redis_port)
+    # TODO add a max retries mechanism, maybe?
```

Comment:
I know this is beyond the scope of this PR but curious how you are thinking about catching FINISHED_WITH_ERROR. Some combination of max retries and checking logs for exceptions or something?

Reply:
I'm still thinking about this one. We need to reproduce a failure and see what we can do here by watching the metrics. At some point it would also be good to have access to the logs for both failures and successful runs as well, maybe have a mechanism to post them in Redis and read it here.

Reply:
Makes sense to me!